File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ProxyCache.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.apache.geode.cache.client.internal;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Properties;
import java.util.Set;

import org.apache.geode.CancelCriterion;
import org.apache.geode.cache.CacheClosedException;
import org.apache.geode.cache.Region;
import org.apache.geode.cache.RegionService;
import org.apache.geode.cache.client.ClientCache;
import org.apache.geode.cache.query.QueryService;
import org.apache.geode.cache.query.internal.ProxyQueryService;
import org.apache.geode.distributed.internal.ServerLocation;
import org.apache.geode.internal.cache.InternalCache;
import org.apache.geode.pdx.JSONFormatter;
import org.apache.geode.pdx.PdxInstance;
import org.apache.geode.pdx.PdxInstanceFactory;
import org.apache.geode.pdx.internal.PdxInstanceFactoryImpl;

/**
 * A wrapper class over an actual Cache instance. This is used when the multiuser-authentication
 * attribute is set to true. Application must use its {@link #getRegion(String)} API instead that of
 * actual Cache instance for getting a reference to Region instances, to perform operations on
 * server.
 *
 * TODO Avoid creating multiple instances of ProxyCache for a single user.
 *
 * @see ClientCache#createAuthenticatedView(Properties)
 * @see ProxyQueryService
 * @see ProxyRegion
 * @since GemFire 6.5
 */
public class ProxyCache implements RegionService {

  /**
   * package-private to avoid synthetic accessor
   * <p>
   * TODO: if this is only in inside client then this should be InternalClientCache
   */
  final InternalCache cache;

  private UserAttributes userAttributes;
  private ProxyQueryService proxyQueryService;
  private boolean isClosed = false;
  private final Stopper stopper = new Stopper();

  public ProxyCache(Properties properties, InternalCache cache, PoolImpl pool) {
    userAttributes = new UserAttributes(properties, pool);
    this.cache = cache;
  }

  @Override
  public void close() {
    close(false);
  }

  public void close(boolean keepAlive) {
    if (isClosed) {
      return;
    }
    // It should go to all the servers it has authenticated itself on and ask
    // them to clean-up its entry from their auth-data structures.
    try {
      if (proxyQueryService != null) {
        proxyQueryService.closeCqs(keepAlive);
      }
      UserAttributes.userAttributes.set(userAttributes);
      for (final ServerLocation serverLocation : userAttributes.getServerToId().keySet()) {
        ProxyCacheCloseOp.executeOn(serverLocation, (ExecutablePool) userAttributes.getPool(),
            userAttributes.getCredentials(), keepAlive);
      }
      List<ProxyCache> proxyCache = ((PoolImpl) userAttributes.getPool()).getProxyCacheList();
      synchronized (proxyCache) {
        proxyCache.remove(this);
      }
    } finally {
      // TODO: I think some NPE will be caused by this code.
      // It would be safer to not null things out.
      // It is really bad that we null out and then set isClosed true.
      isClosed = true;
      proxyQueryService = null;
      userAttributes.setCredentials(null);
      userAttributes = null;
      UserAttributes.userAttributes.set(null);
    }
  }

  @Override
  public QueryService getQueryService() {
    preOp();
    if (proxyQueryService == null) {
      proxyQueryService =
          new ProxyQueryService(this, userAttributes.getPool().getQueryService());
    }
    return proxyQueryService;
  }

  @Override
  public JSONFormatter getJsonFormatter() {
    return new JSONFormatter(this);
  }

  @Override
  public <K, V> Region<K, V> getRegion(String path) {
    preOp();
    if (cache.getRegion(path) == null) {
      return null;
    } else {
      if (!cache.getRegion(path).getAttributes().getDataPolicy().isEmpty()) {
        throw new IllegalStateException(
            "Region's data-policy must be EMPTY when multiuser-authentication is true");
      }
      return new ProxyRegion(this, cache.getRegion(path), cache.getStatisticsClock());
    }
  }

  @Override
  public boolean isClosed() {
    return isClosed;
  }

  public void setProperties(Properties properties) {
    preOp();
    userAttributes.setCredentials(properties);
  }

  public Properties getProperties() {
    preOp();
    return userAttributes.getCredentials();
  }

  public void setUserAttributes(UserAttributes userAttributes) {
    preOp();
    this.userAttributes = userAttributes;
  }

  public UserAttributes getUserAttributes() {
    preOp();
    return userAttributes;
  }

  private void preOp() {
    stopper.checkCancelInProgress(null);
  }

  protected class Stopper extends CancelCriterion {
    @Override
    public String cancelInProgress() {
      String reason = cache.getCancelCriterion().cancelInProgress();
      if (reason != null) {
        return reason;
      }
      if (isClosed()) {
        return "Authenticated cache view is closed for this user.";
      }
      return null;
    }

    @Override
    public RuntimeException generateCancelledException(Throwable e) {
      String reason = cancelInProgress();
      if (reason == null) {
        return null;
      }
      RuntimeException result = cache.getCancelCriterion().generateCancelledException(e);
      if (result != null) {
        return result;
      }
      if (e == null) {
        // Caller did not specify any root cause, so just use our own.
        return cache.getCacheClosedException(reason);
      }

      try {
        return cache.getCacheClosedException(reason, e);
      } catch (IllegalStateException ignore) {
        // Bug 39496 (Jrockit related) Give up. The following
        // error is not entirely sane but gives the correct general picture.
        return new CacheClosedException(reason);
      }
    }
  }

  @Override
  public CancelCriterion getCancelCriterion() {
    return stopper;
  }

  @Override
  public Set<Region<?, ?>> rootRegions() {
    preOp();
    Set<Region<?, ?>> rootRegions = new HashSet<>();
    for (Region<?, ?> region : cache.rootRegions()) {
      if (!region.getAttributes().getDataPolicy().withStorage()) {
        rootRegions.add(new ProxyRegion(this, region, cache.getStatisticsClock()));
      }
    }
    return Collections.unmodifiableSet(rootRegions);
  }

  @Override
  public PdxInstanceFactory createPdxInstanceFactory(String className) {
    return PdxInstanceFactoryImpl.newCreator(className, true, cache);
  }

  public PdxInstanceFactory createPdxInstanceFactory(String className, boolean expectDomainClass) {
    return PdxInstanceFactoryImpl.newCreator(className, expectDomainClass, cache);
  }

  @Override
  public PdxInstance createPdxEnum(String className, String enumName, int enumOrdinal) {
    return PdxInstanceFactoryImpl.createPdxEnum(className, enumName, enumOrdinal, cache);
  }

  /** return a CacheClosedException with the given reason */
  public CacheClosedException getCacheClosedException(String reason) {
    return cache.getCacheClosedException(reason);
  }
}

File: src/cppcache/src/ProxyCache.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <gfcpp/gfcpp_globals.hpp>
#include <gfcpp/SharedPtr.hpp>

#include <gfcpp/DistributedSystem.hpp>
#include "DistributedSystemImpl.hpp"
#include "CacheXmlParser.hpp"
#include "CacheRegionHelper.hpp"
#include <gfcpp/Cache.hpp>
#include "CacheImpl.hpp"
#include "UserAttributes.hpp"
#include "ProxyRegion.hpp"
#include <gfcpp/FunctionService.hpp>
#include "ProxyRemoteQueryService.hpp"
#include "FunctionServiceImpl.hpp"
#include "ProxyCache.hpp"
#include <string.h>
#include <gfcpp/PoolManager.hpp>
#include "ThinClientPoolDM.hpp"
#include "PdxInstanceFactoryImpl.hpp"

using namespace gemfire;

/**
 * Indicates if this cache has been closed.
 * After a new cache object is created, this method returns false;
 * After the close is called on this cache object, this method
 * returns true.
 *
 * @return true, if this cache is closed; false, otherwise
 */
bool ProxyCache::isClosed() const { return m_isProxyCacheClosed; }

/**
 * Terminates this object cache and releases all the local resources.
 * After this cache is closed, any further
 * method call on this cache or any region object will throw
 * <code>CacheClosedException</code>, unless otherwise noted.
 * @param keepalive whether to keep the durable client's queue
 * @throws CacheClosedException,  if the cache is already closed.
 */
void ProxyCache::close() {
  LOGDEBUG("ProxyCache::close: isProxyCacheClosed = %d", m_isProxyCacheClosed);
  if (!m_isProxyCacheClosed) {
    if (m_remoteQueryService != NULLPTR) {
      ProxyRemoteQueryService* prqs =
          static_cast<ProxyRemoteQueryService*>(m_remoteQueryService.ptr());
      prqs->closeCqs(false);
    }

    ProxyCachePtr pcp(this);
    GuardUserAttribures gua(pcp);
    m_isProxyCacheClosed = true;
    m_userAttributes->unSetCredentials();
    // send message to server
    PoolPtr userAttachedPool = m_userAttributes->getPool();
    PoolPtr pool = PoolManager::find(userAttachedPool->getName());
    if (pool != NULLPTR && pool.ptr() == userAttachedPool.ptr()) {
      ThinClientPoolDMPtr poolDM(static_cast<ThinClientPoolDM*>(pool.ptr()));
      if (!poolDM->isDestroyed()) {
        poolDM->sendUserCacheCloseMessage(false);
      }
    }
    return;
  }
  throw IllegalStateException("User cache has been closed.");
}

RegionPtr ProxyCache::getRegion(const char* path) {
  LOGDEBUG("ProxyCache::getRegion:");

  if (!m_isProxyCacheClosed) {
    RegionPtr result;
    CachePtr realCache = CacheFactory::getAnyInstance();

    if (realCache != NULLPTR && !realCache->isClosed()) {
      CacheRegionHelper::getCacheImpl(realCache.ptr())->getRegion(path, result);
    }

    if (result != NULLPTR) {
      PoolPtr userAttachedPool = m_userAttributes->getPool();
      PoolPtr pool = PoolManager::find(result->getAttributes()->getPoolName());
      if (pool != NULLPTR && pool.ptr() == userAttachedPool.ptr() &&
          !pool->isDestroyed()) {
        ProxyRegionPtr pRegion(new ProxyRegion(this, result));
        return pRegion;
      }
      throw IllegalArgumentException(
          "The Region argument is not attached with the pool, which used to "
          "create this user cache.");
    }

    return result;
  }
  throw IllegalStateException("User cache has been closed.");
}

/**
 * Returns a set of root regions in the cache. Does not cause any
 * shared regions to be mapped into the cache. This set is a snapshot and
 * is not backed by the Cache. The regions passed in are cleared.
 *
 * @param regions the region collection object containing the returned set of
 * regions when the function returns
 */

QueryServicePtr ProxyCache::getQueryService() {
  if (!m_isProxyCacheClosed) {
    if (m_remoteQueryService != NULLPTR) return m_remoteQueryService;
    QueryServicePtr prqsPtr(new ProxyRemoteQueryService(this));
    m_remoteQueryService = prqsPtr;
    return prqsPtr;
  }
  throw IllegalStateException("User cache has been closed.");
}

void ProxyCache::rootRegions(VectorOfRegion& regions) {
  LOGDEBUG("ProxyCache::rootRegions:");

  if (!m_isProxyCacheClosed) {
    RegionPtr result;
    CachePtr realCache = CacheFactory::getAnyInstance();

    if (realCache != NULLPTR && !realCache->isClosed()) {
      VectorOfRegion tmp;
      // this can cause issue when pool attached with region in multiuserSecure
      // mode
      realCache->rootRegions(tmp);

      if (tmp.size() > 0) {
        for (int32_t i = 0; i < tmp.size(); i++) {
          RegionPtr reg = tmp.at(i);
          if (strcmp(m_userAttributes->getPool()->getName(),
                     reg->getAttributes()->getPoolName()) == 0) {
            ProxyRegionPtr pRegion(new ProxyRegion(this, reg));
            regions.push_back(pRegion);
          }
        }
      }
    }
  }
}

ProxyCache::ProxyCache(PropertiesPtr credentials, PoolPtr pool) {
  m_remoteQueryService = NULLPTR;
  m_isProxyCacheClosed = false;
  UserAttributesPtr userAttr;
  userAttr = new UserAttributes(credentials, pool, this);
  m_userAttributes = userAttr;
}

ProxyCache::~ProxyCache() {}

PdxInstanceFactoryPtr ProxyCache::createPdxInstanceFactory(
    const char* className) {
  PdxInstanceFactoryPtr pIFPtr(new PdxInstanceFactoryImpl(className));
  return pIFPtr;
}
